va/vampic.go

```diff
@@ -0,0 +1,427 @@
+	if core.IsAnyNilOrZero(req, req.DnsName, req.Challenge, req.Authz, req.ExpectedKeyAuthorization) {
+		return nil, berrors.InternalServerError("Incomplete validation request")
+	}
+
+	chall, err := bgrpc.PBToChallenge(req.Challenge)
+	if err != nil {
+		return nil, errors.New("challenge failed to deserialize")
+	}
+
+	err = chall.CheckPending()
+	if err != nil {
+		return nil, berrors.MalformedError("challenge failed consistency check: %s", err)
+	}
+
+	// Set up variables and a deferred closure to report validation latency
+	// metrics and log validation errors. Below here, do not use := to redeclare
+	// `prob`, or this will fail.
+	var prob *probs.ProblemDetails
+	var summary *mpicSummary
+	var localLatency time.Duration
+	start := va.clk.Now()
+	logEvent := validationLogEvent{
+		AuthzID:    req.Authz.Id,
+		Requester:  req.Authz.RegID,
+		Identifier: req.DnsName,
+		Challenge:  chall,
+	}
+	defer func() {
+		probType := ""
+		outcome := fail
+		if prob != nil {
+			probType = string(prob.Type)
+			logEvent.Error = prob.Error()
+			logEvent.Challenge.Error = prob
+			logEvent.Challenge.Status = core.StatusInvalid
+		} else {
+			logEvent.Challenge.Status = core.StatusValid
+			outcome = pass
+		}
+		// Observe local validation latency (primary|remote).
+		va.observeLatency(opChall, va.perspective, string(chall.Type), probType, outcome, localLatency)
+		if va.isPrimaryVA() {
+			// Observe total validation latency (primary+remote).
+			va.observeLatency(opChall, allPerspectives, string(chall.Type), probType, outcome, va.clk.Since(start))
+			logEvent.Summary = summary
+		}
+
+		// Log the total validation latency.
+		logEvent.Latency = va.clk.Since(start).Round(time.Millisecond).Seconds()
+		va.log.AuditObject("Validation result", logEvent)
+	}()
+
+	// Do local validation. Note that we process the result in a couple ways
+	// *before* checking whether it returned an error. These few checks are
+	// carefully written to ensure that they work whether the local validation
+	// was successful or not, and cannot themselves fail.
+	records, err := va.validateChallenge(
+		ctx,
+		identifier.NewDNS(req.DnsName),
+		chall.Type,
+		chall.Token,
+		req.ExpectedKeyAuthorization,
+	)
+
+	// Stop the clock for local validation latency.
+	localLatency = va.clk.Since(start)
+
+	// Check for malformed ValidationRecords
+	logEvent.Challenge.ValidationRecord = records
+	if err == nil && !logEvent.Challenge.RecordsSane() {
+		err = errors.New("records from local validation failed sanity check")
+	}
+
+	if err != nil {
+		logEvent.InternalError = err.Error()
+		prob = detailedError(err)
+		return bgrpc.ValidationResultToPB(records, filterProblemDetails(prob), va.perspective, va.rir)
+	}
+
+	if va.isPrimaryVA() {
+		// Do remote validation. We do this after local validation is complete to
+		// avoid wasting work when validation will fail anyway. This only returns a
+		// singular problem, because the remote VAs have already audit-logged their
```

Comment:
I wouldn't phrase it this way: the remote VAs have already _logged_ their problems, so we can debug them and don't need to re-log them here. But they aren't audit-logged in the way that the BRs require us to log various MPIC things, so I'd maybe stay away from saying "audit-logged".

_edit_: I now see that this is a copy-paste and only shows up as a diff due to indentation whitespace, but I'd maybe like to change it anyway.

Reply:
Done.